executor/ddl.go

```diff
@@ -287,3 +291,160 @@ func (e *DDLExec) executeAlterTable(s *ast.AlterTableStmt) error {
+	}
+	// Call DDL RestoreTable
+	err = domain.GetDomain(e.ctx).DDL().RestoreTable(e.ctx, tblInfo, job.SchemaID, autoID, job.ID, job.StartTS)
+	return errors.Trace(err)
+}
+
+func getRestoreTableByJobID(e *RestoreTableExec, t *meta.Meta, dom *domain.Domain) (*model.Job, *model.TableInfo, error) {
+	job, err := t.GetHistoryDDLJob(e.jobID)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	if job == nil {
+		return nil, nil, admin.ErrDDLJobNotFound.GenWithStackByArgs(e.jobID)
+	}
+	if job.Type != model.ActionDropTable {
+		return nil, nil, errors.Errorf("Job %v type is %v, not drop table", job.ID, job.Type)
+	}
+
+	// Check gc safe point for getting snapshot infoSchema.
+	err = gcutil.ValidateSnapshot(e.ctx, job.StartTS)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	// Get the snapshot infoSchema before drop table.
+	snapInfo, err := dom.GetSnapshotInfoSchema(job.StartTS)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	// Get table meta from snapshot infoSchema.
+	table, ok := snapInfo.TableByID(job.TableID)
+	if !ok {
+		return nil, nil, infoschema.ErrTableNotExists.GenWithStackByArgs(
+			fmt.Sprintf("(Schema ID %d)", job.SchemaID),
+			fmt.Sprintf("(Table ID %d)", job.TableID),
+		)
+	}
+	return job, table.Meta(), nil
+}
+
+func getRestoreTableByTableName(e *RestoreTableExec, t *meta.Meta, dom *domain.Domain) (*model.Job, *model.TableInfo, error) {
+	jobs, err := t.GetAllHistoryDDLJobs()
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	var job *model.Job
+	var tblInfo *model.TableInfo
+	gcSafePoint, err := gcutil.GetGCSafePoint(e.ctx)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	schemaName := e.Table.Schema.L
+	if schemaName == "" {
+		schemaName = e.ctx.GetSessionVars().CurrentDB
+	}
+	if schemaName == "" {
+		return nil, nil, errors.Trace(core.ErrNoDB)
+	}
+	// TODO: only search recent `e.JobNum` ddl jobs.
+	for i := len(jobs) - 1; i > 0; i-- {
+		job = jobs[i]
+		if job.Type != model.ActionDropTable {
+			continue
+		}
+		// Check gc safe point for getting snapshot infoSchema.
+		err = gcutil.ValidateSnapshotWithGCSafePoint(job.StartTS, gcSafePoint)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		// Get the snapshot infoSchema before drop table.
+		snapInfo, err := dom.GetSnapshotInfoSchema(job.StartTS)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		// Get table meta from snapshot infoSchema.
+		table, ok := snapInfo.TableByID(job.TableID)
+		if !ok {
+			return nil, nil, infoschema.ErrTableNotExists.GenWithStackByArgs(
+				fmt.Sprintf("(Schema ID %d)", job.SchemaID),
+				fmt.Sprintf("(Table ID %d)", job.TableID),
+			)
+		}
+		if table.Meta().Name.L == e.Table.Name.L {
+			schema, ok := dom.InfoSchema().SchemaByID(job.SchemaID)
+			if !ok {
+				return nil, nil, errors.Trace(infoschema.ErrDatabaseNotExists.GenWithStackByArgs(
+					fmt.Sprintf("(Schema ID %d)", job.SchemaID),
+				))
+			}
+			if schema.Name.L == schemaName {
```

Comment:
Why do we need to check this? The` schemaID` is unique.

Reply:
Here I want to make sure whether is this job drops the table.
think about this: 2 databases have the same table name: `db1.t, db2.t`
```sql
drop table db1.t;
drop table db2.t;
admin restore table db1.t; // this will restore db2.t, but not db1.t.
```

Reply:
Yes, But here I want to make sure the database of the job is the same database of the job. here you still can not make sure that job drop the table which you want to restore.

Reply:
Got it.